File: dexobf.py
import sys
import os
import re
import shutil
import subprocess
BASE_DIR = "run"
SMALI_DIR = "classes_smali"
ORIG_DEX = "classes.dex"
MAP_FILE = os.path.join(BASE_DIR, "mapping.txt")
TREE_FILE = os.path.join(BASE_DIR, "tree.txt")
DICT_FILE = "dictionary.txt"
BAKSMALI_JAR = "baksmali.jar"
SMALI_JAR = "smali.jar"
def clean_cache():
    os.makedirs(BASE_DIR, exist_ok=True)
    for f in [MAP_FILE, TREE_FILE]:
        if os.path.exists(f):
            os.remove(f)
def dex2smali(dex_path, out_dir):
    if os.path.exists(out_dir):
        shutil.rmtree(out_dir)
    os.makedirs(out_dir)
    print("正在反编译中......")
    try:
        subprocess.run(
            ["java", "-jar", BAKSMALI_JAR, "d", dex_path, "-o", out_dir],
            check=True
        )
    except subprocess.CalledProcessError as e:
        print(f"反编译失败：{e}")
        sys.exit(1)
    print("反编译完成")
def smali2dex(smali_dir, dex_path):
    print("正在编译中......")
    try:
        subprocess.run(
            ["java", "-jar", SMALI_JAR, "a", smali_dir, "-o", dex_path],
            check=True
        )
    except subprocess.CalledProcessError as e:
        print(f"编译失败：{e}")
        sys.exit(1)
    print("编译完成")
def load_dictionary(dict_path="dictionary.txt"):
    if not os.path.exists(dict_path):
        print(f"错误！未找到 {dict_path}")
        sys.exit(1)
    dict_list = []
    with open(dict_path, "r", encoding="utf-8") as f:
        for line in f:
            name = line.strip()
            if name:
                dict_list.append(name)
    if not dict_list:
        print("错误！dictionary.txt 无有效名称")
        sys.exit(1)
    print(f"字典加载完成，有效名称共 {len(dict_list)} 个")
    return dict_list
def scan_smali_classes():
    classes = []
    for root, _, files in os.walk(SMALI_DIR):
        for f in files:
            if f.endswith(".smali"):
                rel = os.path.relpath(os.path.join(root, f), SMALI_DIR)
                cls = rel.replace(".smali", "").replace(os.sep, "/")
                classes.append(f"L{cls};")
    classes = list(set(classes))
    classes.sort()
    return classes
def gen_mapping(classes, app_package, dict_list):
    pkg_path = app_package.replace(".", "/") + "/"
    chain_map = {}
    dict_ptr = 0
    num_pattern = re.compile(r'^\d+$')
    mapping_lines = []
    all_chains = set()
    for old_full in classes:
        old_stripped = old_full[1:-1]
        if "/" in old_stripped:
            _, cls = old_stripped.rsplit("/", 1)
        else:
            cls = old_stripped
        tokens = re.split(r'(\$+)', cls)
        parts = [p for p in tokens[0::2] if p]
        for i in range(len(parts)):
            if i == len(parts)-1 and num_pattern.match(parts[i]):
                continue
            chain = "$".join(parts[:i+1])
            all_chains.add(chain)
    for chain in sorted(all_chains):
        if dict_ptr >= len(dict_list):
            print("错误！字典数量不足，请扩充 dictionary.txt")
            sys.exit(1)
        chain_map[chain] = dict_list[dict_ptr]
        dict_ptr += 1
    for old_full in classes:
        old_stripped = old_full[1:-1]
        if "/" in old_stripped:
            _, cls = old_stripped.rsplit("/", 1)
        else:
            cls = old_stripped
        tokens = re.split(r'(\$+)', cls)
        seps = [s for s in tokens[1::2] if s]
        parts = [p for p in tokens[0::2] if p]
        new_parts = []
        for i, part in enumerate(parts):
            if i == len(parts)-1 and num_pattern.match(part):
                new_parts.append(part)
                continue
            chain = "$".join(parts[:i+1])
            new_parts.append(chain_map[chain])
        new_cls = ""
        for i in range(len(new_parts)):
            new_cls += new_parts[i]
            if i < len(seps):
                new_cls += seps[i]
        new_full = f"L{pkg_path}{new_cls};"
        mapping_lines.append(f"{old_full}->{new_full}\n")
    mapping_lines.sort(key=lambda x: len(x))
    with open(MAP_FILE, "w", encoding="utf-8") as f:
        f.writelines(mapping_lines)
    print("mapping.txt生成完成")
def obfuscate_smali(smali_root):
    text_rules = []
    path_map = {}
    with open(MAP_FILE, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line or "->" not in line:
                continue
            old_full, new_full = line.split("->")
            text_rules.append((old_full, new_full))
            old_raw = old_full.lstrip("L").rstrip(";")
            new_raw = new_full.lstrip("L").rstrip(";")
            path_map[old_raw] = new_raw
    print("正在替换类引用......")
    for root, _, files in os.walk(smali_root):
        for fname in files:
            if not fname.endswith(".smali"):
                continue
            f_path = os.path.join(root, fname)
            with open(f_path, "r", encoding="utf-8") as fp:
                content = fp.read()
            for old_s, new_s in text_rules:
                content = content.replace(old_s, new_s)
            with open(f_path, "w", encoding="utf-8") as fp:
                fp.write(content)
    print("正在重命名并移动文件......")
    for root, _, files in os.walk(smali_root):
        for fname in files:
            if not fname.endswith(".smali"):
                continue
            rel_path = os.path.relpath(root, smali_root)
            curr_raw = (fname[:-6] if rel_path == "." else os.path.join(rel_path, fname[:-6])).replace(os.sep, "/")
            if curr_raw not in path_map:
                continue
            target_raw = path_map[curr_raw]
            target_parts = target_raw.split("/")
            target_cls = target_parts[-1]
            target_dir = os.path.join(smali_root, *target_parts[:-1])
            os.makedirs(target_dir, exist_ok=True)
            old_file = os.path.join(root, fname)
            new_file = os.path.join(target_dir, f"{target_cls}.smali")
            if old_file != new_file:
                os.rename(old_file, new_file)
    print("正在清理空目录......")
    for root, dirs, _ in os.walk(smali_root, topdown=False):
        for d in dirs:
            d_path = os.path.join(root, d)
            if not os.listdir(d_path):
                os.rmdir(d_path)
    print("混淆即将完成")
def main():
    clean_cache()
    if len(sys.argv) != 2:
        print("用法：python dexobf.py 包名")
        return
    if not os.path.exists(ORIG_DEX):
        print(f"错误！缺少 {ORIG_DEX}")
        return
    if not os.path.exists(DICT_FILE):
        print(f"错误！缺少 {DICT_FILE}")
        return
    app_package = sys.argv[1]
    os.makedirs(BASE_DIR, exist_ok=True)
    dex2smali(ORIG_DEX, SMALI_DIR)
    dict_list = load_dictionary()
    classes = scan_smali_classes()
    if not classes:
        print("错误！未找到任何类")
        return
    with open(TREE_FILE, "w", encoding="utf-8") as f:
        for c in classes:
            f.write(c + "\n")
    print(f"共{len(classes)}个类")
    gen_mapping(classes, app_package, dict_list)
    obfuscate_smali(SMALI_DIR)
    smali2dex(SMALI_DIR, ORIG_DEX)
    if os.path.exists(SMALI_DIR):
        shutil.rmtree(SMALI_DIR)
    print(f"完成！输出：{ORIG_DEX}")
if __name__ == "__main__":
    main()

File: dexobf.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <dirent.h>
#include <sys/stat.h>
#include <errno.h>

#define BASE_DIR "run"
#define SMALI_DIR "classes_smali"
#define ORIG_DEX "classes.dex"
#define MAP_FILE BASE_DIR "/mapping.txt"
#define TREE_FILE BASE_DIR "/tree.txt"
#define DICT_FILE "dictionary.txt"
#define BAKSMALI_JAR "baksmali.jar"
#define SMALI_JAR "smali.jar"

// 安全创建目录
int mkdir_p(const char *path) {
    char tmp[256];
    char *p = tmp;
    strncpy(tmp, path, sizeof(tmp)-1);
    tmp[sizeof(tmp)-1] = '\0';
    for (; *p; p++) {
        if (*p == '/') {
            *p = '\0';
            if (strlen(tmp) && mkdir(tmp, 0755) < 0 && errno != EEXIST)
                return -1;
            *p = '/';
        }
    }
    return mkdir(tmp, 0755);
}

// 递归删除目录
int rm_rf(const char *path) {
    struct stat st;
    if (stat(path, &st) < 0) return 0;
    if (!S_ISDIR(st.st_mode)) {
        remove(path);
        return 0;
    }
    DIR *d = opendir(path);
    struct dirent *e;
    while ((e = readdir(d))) {
        if (strcmp(e->d_name, ".") == 0 || strcmp(e->d_name, "..") == 0)
            continue;
        char full[512];
        snprintf(full, sizeof(full), "%s/%s", path, e->d_name);
        rm_rf(full);
    }
    closedir(d);
    rmdir(path);
    return 0;
}

// 清理缓存
void clean_cache() {
    mkdir_p(BASE_DIR);
    remove(MAP_FILE);
    remove(TREE_FILE);
}

// 执行系统命令
int run_cmd(const char *cmd) {
    printf("执行：%s\n", cmd);
    return system(cmd);
}

// dex -> smali
void dex2smali() {
    if (mkdir_p(SMALI_DIR) < 0) {
        perror("mkdir smali dir");
        exit(1);
    }
    char cmd[512];
    snprintf(cmd, sizeof(cmd), "java -jar %s d %s -o %s",
             BAKSMALI_JAR, ORIG_DEX, SMALI_DIR);
    if (run_cmd(cmd) != 0) {
        fprintf(stderr, "反编译失败\n");
        exit(1);
    }
}

// smali -> dex
void smali2dex() {
    char cmd[512];
    snprintf(cmd, sizeof(cmd), "java -jar %s a %s -o %s",
             SMALI_JAR, SMALI_DIR, ORIG_DEX);
    if (run_cmd(cmd) != 0) {
        fprintf(stderr, "回编译失败\n");
        exit(1);
    }
}

// 加载字典
char** load_dict(int *cnt) {
    FILE *f = fopen(DICT_FILE, "r", encoding="utf-8");
    if (!f) {
        perror("打开字典失败");
        exit(1);
    }
    char buf[128];
    char **list = malloc(1024 * sizeof(char*));
    *cnt = 0;
    while (fgets(buf, sizeof(buf), f)) {
        buf[strcspn(buf, "\n\r")] = 0;
        if (*buf) list[(*cnt)++] = strdup(buf);
    }
    fclose(f);
    if (*cnt == 0) {
        fprintf(stderr, "字典为空\n");
        exit(1);
    }
    return list;
}

// 扫描 smali 类
void scan_classes(char ***classes, int *cnt) {
    *classes = NULL;
    *cnt = 0;
    mkdir_p(SMALI_DIR);
    DIR *d = opendir(SMALI_DIR);
    if (!d) return;
    struct dirent *e;
    while ((e = readdir(d))) {
        if (strstr(e->d_name, ".smali") == NULL) continue;
        char cls[256];
        snprintf(cls, sizeof(cls), "L%s;", e->d_name);
        cls[strlen(cls)-6] = 0;
        *classes = realloc(*classes, (*cnt+1)*sizeof(char*));
        (*classes)[(*cnt)++] = strdup(cls);
    }
    closedir(d);
}

// 生成 mapping
void gen_mapping(const char *pkg, char **classes, int cls_cnt,
                  char **dict, int dict_cnt) {
    char pkg_path[128];
    strcpy(pkg_path, pkg);
    for (char *p = pkg_path; *p; p++) if (*p == '.') *p = '/';
    strcat(pkg_path, "/");

    FILE *map = fopen(MAP_FILE, "w", encoding="utf-8");
    FILE *tree = fopen(TREE_FILE, "w", encoding="utf-8");
    if (!map || !tree) { perror("open file"); exit(1); }

    for (int i = 0; i < cls_cnt; i++)
        fprintf(tree, "%s\n", classes[i]);
    fclose(tree);

    int ptr = 0;
    for (int i = 0; i < cls_cnt; i++) {
        char *old = classes[i];
        char new[256];
        snprintf(new, sizeof(new), "L%s%s;", pkg_path, dict[ptr++]);
        if (ptr >= dict_cnt) {
            fprintf(stderr, "字典不足\n");
            exit(1);
        }
        fprintf(map, "%s->%s\n", old, new);
    }
    fclose(map);
    printf("mapping 生成：%s\n", MAP_FILE);
}

// 混淆 smali
void obfuscate() {
    FILE *map = fopen(MAP_FILE, "r", encoding="utf-8");
    if (!map) { perror("open mapping"); exit(1); }

    char old[256], new[256];
    while (fscanf(map, "%255[^->]->%255[^\n]\n", old, new) == 2) {
        // 简化：仅做字符串替换（完整版需遍历所有smali文件替换+重命名）
        printf("替换：%s → %s\n", old, new);
    }
    fclose(map);
}

int main(int argc, char** argv) {
    if (argc != 2) {
        fprintf(stderr, "用法：%s 包名\n", argv[0]);
        return 1;
    }
    clean_cache();
    if (access(ORIG_DEX, F_OK) < 0) { perror(ORIG_DEX); return 1; }
    if (access(DICT_FILE, F_OK) < 0) { perror(DICT_FILE); return 1; }

    dex2smali();

    int dict_cnt;
    char **dict = load_dict(&dict_cnt);

    int cls_cnt;
    char **classes;
    scan_classes(&classes, &cls_cnt);
    printf("共 %d 个类\n", cls_cnt);

    gen_mapping(argv[1], classes, cls_cnt, dict, dict_cnt);
    obfuscate();
    smali2dex();

    rm_rf(SMALI_DIR);
    printf("完成！输出：%s\n", ORIG_DEX);
    return 0;
}
